### PR TITLE
Add action and dom event instrumentation

### PR DESCRIPTION
**Pull request by asakusuma**
This is just the action and dom instrumentation part copied and merged over from @chadhietala's PR here: https://github.com/emberjs/ember.js/pull/13017

This is the highest priority part for getting us off of private APIs for tracking.

I noticed that some of the tests seem to be disabled for glimmer2, so I'm wondering what I need to do to get this working and tested with glimmer2.

cc @krisselden

### CODE REVIEW

File: features.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
     "ember-metal-ember-assign": true,
     "ember-htmlbars-local-lookup": true,
     "ember-application-engines": null,
-    "ember-glimmer": null
+    "ember-glimmer": null,
+    "ember-improved-instrumentation": true
```
Comment (mixonic) [inline features.json]: This should be `null` in the PR.

File: packages/ember-metal/lib/instrumentation.js [hunk 2]
```diff
@@ -105,6 +106,38 @@ export function instrument(name, _payload, callback, binding) {
...
+  to call the finalizer that this function returns at some point
+  in the future.
+  @public
+  @method asyncInstrument
+  @param {String} [name] Namespaced event name
+  @param {Function} _payload
+  @return {Function} a finalizer function to call when the instrumented
+    work has finished.
+*/
+export function asyncInstrument(name, _payload) {
```
Comment (mixonic) [inline packages/ember-metal/lib/instrumentation.js]: Where is this used?

File: packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js [hunk 3]
```diff
@@ -583,3 +584,38 @@ QUnit.test('objects that define INVOKE can be casted to actions', function(asser
...
 
 }
+
+if (isEnabled('ember-improved-instrumentation')) {
+  QUnit.test('action should fire interaction event', function(assert) {
+    assert.expect(2);
+
+    subscribe('interaction.ember-action', {
+      before() {
+        console.log(arguments);
```
Comment (mixonic) [inline packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js]: `console.log`

File: packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js [hunk 4]
```diff
@@ -583,3 +584,38 @@ QUnit.test('objects that define INVOKE can be casted to actions', function(asser
...
+    });
+
+    innerComponent = EmberComponent.extend({
+      fireAction() {
+        this.attrs.submit();
+      }
+    }).create();
+
+    outerComponent = EmberComponent.extend({
+      layout: compile('{{view innerComponent submit=(action outerSubmit)}}'),
```
Comment (mixonic) [inline packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js]: Would be nice to not use the `view` helper in new tests.

File: packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js [hunk 5]
```diff
@@ -583,3 +584,38 @@ QUnit.test('objects that define INVOKE can be casted to actions', function(asser
...
+  QUnit.test('action should fire interaction event', function(assert) {
+    assert.expect(2);
+
+    subscribe('interaction.ember-action', {
+      before() {
+        console.log(arguments);
+        assert.ok(true, 'instrumentation subscriber was called');
+      },
+      after() {
+      }
```
Comment (mixonic) [inline packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js]: Empty `after() {` can probably be dropped.

Comment (asakusuma) [inline packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js]: Removed

File: packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js (hunk 4 above)
Comment (asakusuma) [inline packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js]: I'm assuming the right thing to use would be a component helper, though I'm unsure the best way to do a component setup in this test. Should I try to use [RenderingTest](https://github.com/emberjs/ember.js/blob/master/packages/ember-glimmer/tests/integration/components/dynamic-components-test.js#L7)? Or perhaps do some custom setup like in [local-lookup-test.js](https://github.com/emberjs/ember.js/blob/master/packages/ember-htmlbars/tests/integration/local-lookup-test.js)?

Comment (chadhietala) [inline packages/ember-routing-htmlbars/tests/helpers/closure_action_test.js]: I think you would need to introduce the `owner` stuff that is in the local-look-test. Basically you need a means of registering.

**Pull request merged**